common/dllserver/dllserver.cpp

```diff
@@ -759,37 +642,135 @@ void DllServer::removeDll(const char * name, bool removeDlls, bool removeDirecto
 
 void DllServer::registerDll(const char * name, const char * kind, const char * dllLocation)
 {
-    doRegisterDll(name, kind, dllLocation, NULL);
+    doRegisterDll(name, kind, dllLocation);
+}
+
+//---------------------------------------------------------------------------
+
+class SharedVolumeDllServer : public DllServerBase
+{
+public:
+    SharedVolumeDllServer(const char * _mountPath) : DllServerBase(_mountPath) {};
+
+    virtual void ensureAvailable(const char * name, DllLocationType location) override;
+    virtual void registerDll(const char * name, const char * kind, const char * dllPath) override;
+};
+
+void SharedVolumeDllServer::ensureAvailable(const char * name, DllLocationType location)
+{
+    Owned<DllEntry> match = doGetEntry(name);
+    if (!match)
+        throwError1(DSVERR_CouldNotFindDll, name);
+    // MORE - if we still want to have roxie copy to directory, may need code here
+#ifdef _DEBUG
+    Owned<IDllLocation> bestLocation = match->getBestLocation();
+    assertex (bestLocation->queryLocation() >= location);
+#endif
+}
+
+void SharedVolumeDllServer::registerDll(const char * name, const char * kind, const char * dllPath)
+{
+    RemoteFilename dllRemote;
+    dllRemote.setRemotePath(dllPath);
+    assertex(dllRemote.isLocal());
+    StringBuffer sourceLocalPath;
+
+    dllRemote.getLocalPath(sourceLocalPath);
+
+    StringBuffer mountPath(rootDir);
+    addPathSepChar(mountPath);
+    splitFilename(sourceLocalPath.str(),NULL,NULL,&mountPath,&mountPath);
+
+    // Copy file to mountpath, if not already there
+    OwnedIFile source = createIFile(dllRemote);
+    OwnedIFile target = createIFile(mountPath);
+    if (!target->exists())
+    {
+        source->copyTo(target, 0, NULL, true);
+    }
+
+    Owned<IRemoteConnection> conn = getEntryConnection(name, RTM_LOCK_WRITE);
+    if (conn)
+    {
+        /* check the entry doesn't exist already....
+         * Ideally the connection above would be a RTM_LOCK_READ and be changed to a RTM_LOCK_WRITE only when 'location' not found
```

Comment:
That should be possible by calling conn->changeMode(), but are you likely to ever match an item that is alrady there?

Reply:
No. (This code is just copied from the legacy case)